File: app/src/main/java/com/kidozh/discuzhub/services/DiscuzApiService.java
package com.kidozh.discuzhub.services;

import com.kidozh.discuzhub.results.BBSIndexResult;
import com.kidozh.discuzhub.results.BaseResult;
import com.kidozh.discuzhub.results.BuyThreadResult;
import com.kidozh.discuzhub.results.DisplayThreadsResult;
import com.kidozh.discuzhub.results.FavoriteForumResult;
import com.kidozh.discuzhub.results.ApiMessageActionResult;
import com.kidozh.discuzhub.results.FavoriteThreadResult;
import com.kidozh.discuzhub.results.ForumResult;
import com.kidozh.discuzhub.results.HotForumsResult;
import com.kidozh.discuzhub.results.LoginResult;
import com.kidozh.discuzhub.results.SecureInfoResult;
import com.kidozh.discuzhub.results.ThreadResult;
import com.kidozh.discuzhub.results.UserNoteListResult;
import com.kidozh.discuzhub.results.VariableResults;

import java.util.HashMap;

import retrofit2.Call;
import retrofit2.http.Field;
import retrofit2.http.FormUrlEncoded;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Query;
import retrofit2.http.QueryMap;

public interface DiscuzApiService {
    public static String DISCUZ_API_PATH = "/api/mobile/index.php";

    @GET(DISCUZ_API_PATH+"?version=4&module=myfavthread")
    Call<FavoriteThreadResult> getFavoriteThreadResult(@Query("page") int page);

    @GET(DISCUZ_API_PATH+"?version=4&module=myfavforum")
    Call<FavoriteForumResult> getFavoriteForumResult(@Query("page") int page);

    @FormUrlEncoded
    @POST(DISCUZ_API_PATH+"?version=4&module=favthread&type=thread&ac=favorite&favoritesubmit=true")
    Call<ApiMessageActionResult> favoriteThreadActionResult(
            @Query("formhash") String formhash,
            @Query("id") int tid,
            @Field("description") String description


    );

    @FormUrlEncoded
    @POST(DISCUZ_API_PATH+"?version=4&module=favthread&type=all&ac=favorite&op=delete&inajax=1&favoritesubmit=true")
    Call<ApiMessageActionResult> unfavoriteThreadActionResult(
            @Field("formhash") String formhash,
            @Field("deletesubmit") String submit,
            @Field("handlekey") String handleKey,
            @Query("favid") int favid


    );

    @FormUrlEncoded
    @POST(DISCUZ_API_PATH+"?version=4&module=favforum&type=forum&ac=favorite&favoritesubmit=true")
    Call<ApiMessageActionResult> favoriteForumActionResult(
            @Query("hash") String formhash,
            @Query("id") int tid,
            @Field("description") String description


    );

    @FormUrlEncoded
    @POST(DISCUZ_API_PATH+"?version=4&module=favforum&type=all&ac=favorite&op=delete&inajax=1&favoritesubmit=true")
    Call<ApiMessageActionResult> unfavoriteForumActionResult(
            @Field("hash") String formhash,
            @Field("deletesubmit") String submit,
            @Field("handlekey") String handleKey,
            @Query("favid") int favid


    );


    @GET(DISCUZ_API_PATH+"?version=4&module=mobilesign")
    Call<ApiMessageActionResult> mobileSignActionResult(
            @Query("hash") String formHash
    );

    @GET(DISCUZ_API_PATH+"?version=4&module=mynotelist")
    Call<UserNoteListResult> userNotificationListResult(
            @Query("page") int page
    );

    @GET(DISCUZ_API_PATH+"?version=4&module=login&mod=logging&action=login")
    Call<LoginResult> getLoginResult();

    @GET(DISCUZ_API_PATH+"?version=4&module=hotforum")
    Call<HotForumsResult> hotForumResult();

    @GET(DISCUZ_API_PATH+"?version=4&module=hotthread")
    Call<DisplayThreadsResult> hotThreadResult(
            @Query("page") int page
    );

    @GET(DISCUZ_API_PATH+"?version=4&module=forumindex")
    Call<BBSIndexResult> indexResult();

    @GET(DISCUZ_API_PATH+"?version=4&module=forumdisplay")
    Call<ForumResult> forumDisplayResult(@QueryMap HashMap<String,String> options);

    @GET(DISCUZ_API_PATH+"?version=4&module=viewthread")
    Call<ThreadResult> viewThreadResult(@QueryMap HashMap<String,String> options);

    @GET(DISCUZ_API_PATH+"?version=4&module=secure")
    Call<SecureInfoResult> secureResult(@Query("type") String type);

    @GET(DISCUZ_API_PATH+"?version=4&module=recommend")
    Call<ApiMessageActionResult> recommendThread(
            @Query("hash") String formhash,
            @Query("tid") int tid
    );

    @GET(DISCUZ_API_PATH+"?version=4&module=recommend&do=substract")
    Call<ApiMessageActionResult> unrecommendThread(
            @Query("hash") String formhash,
            @Query("tid") int tid
    );

    @GET(DISCUZ_API_PATH+"?version=4&module=buythread")
    Call<BuyThreadResult> getThreadPriceInfo(
            @Query("tid") int tid
    );

    @FormUrlEncoded
    @POST(DISCUZ_API_PATH+"?version=4&module=buythread&paysubmit=yes")
    Call<BuyThreadResult> buyThread(
            @Field("tid") int tid,
            @Field("formhash") String formhash,
            @Query("handlekey") String pay
    );
}

File: app/src/main/java/com/kidozh/discuzhub/services/DiscuzApiService.kt
package com.kidozh.discuzhub.services

import com.kidozh.discuzhub.results.*
import retrofit2.Call
import retrofit2.http.*
import java.util.*

interface DiscuzApiService {
    @GET("$DISCUZ_API_PATH?version=4&module=myfavthread")
    fun getFavoriteThreadResult(@Query("page") page: Int): Call<FavoriteThreadResult>

    @GET("$DISCUZ_API_PATH?version=4&module=myfavforum")
    fun getFavoriteForumResult(@Query("page") page: Int): Call<FavoriteForumResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=favthread&type=thread&ac=favorite&favoritesubmit=true")
    fun favoriteThreadActionResult(
        @Query("formhash") formhash: String,
        @Query("id") tid: Int,
        @Field("description") description: String
    ): Call<ApiMessageActionResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=favthread&type=all&ac=favorite&op=delete&inajax=1&favoritesubmit=true")
    fun unfavoriteThreadActionResult(
        @Field("formhash") formhash: String,
        @Field("deletesubmit") submit: String,
        @Field("handlekey") handleKey: String,
        @Query("favid") favid: Int
    ): Call<ApiMessageActionResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=favforum&type=forum&ac=favorite&favoritesubmit=true")
    fun favoriteForumActionResult(
        @Query("formhash") formhash: String,
        @Query("id") tid: Int,
        @Field("description") description: String
    ): Call<ApiMessageActionResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=favforum&type=all&ac=favorite&op=delete&inajax=1&favoritesubmit=true")
    fun unfavoriteForumActionResult(
        @Field("formhash") formhash: String,
        @Field("deletesubmit") submit: String,
        @Field("handlekey") handleKey: String,
        @Query("favid") favid: Int
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=mobilesign")
    fun mobileSignActionResult(
        @Query("hash") formHash: String
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=mynotelist")
    fun userNotificationListResult(
        @Query("page") page: Int
    ): Call<UserNoteListResult>

    @get:GET("$DISCUZ_API_PATH?version=4&module=login&mod=logging&action=login")
    val loginResult: Call<LoginResult>

    @GET("$DISCUZ_API_PATH?version=4&module=hotforum")
    fun hotForumResult(): Call<HotForumsResult>

    @GET("$DISCUZ_API_PATH?version=4&module=hotthread")
    fun hotThreadResult(
        @Query("page") page: Int
    ): Call<DisplayThreadsResult>

    @GET("$DISCUZ_API_PATH?version=4&module=forumindex")
    fun indexResult(): Call<DiscuzIndexResult>

    @GET("$DISCUZ_API_PATH?version=4&module=forumindex")
    fun indexResultRaw(): Call<String>

    @GET("$DISCUZ_API_PATH?version=4&module=forumdisplay")
    fun forumDisplayResult(@QueryMap options: HashMap<String, String>): Call<ForumResult>

    @GET("$DISCUZ_API_PATH?version=4&module=viewthread")
    fun viewThreadResult(@QueryMap options: HashMap<String, String>): Call<ThreadResult>

    @GET("$DISCUZ_API_PATH?version=4&module=viewthread")
    fun viewThreadResultRaw(@QueryMap options: HashMap<String, String>): Call<String>

    @GET("$DISCUZ_API_PATH?version=4&module=secure")
    fun secureResult(@Query("type") type: String): Call<SecureInfoResult>

    @GET("$DISCUZ_API_PATH?version=4&module=secure")
    fun captchaCall(@Query("type") type: String): Call<SecureInfoResult>

    @GET("$DISCUZ_API_PATH?version=4&module=recommend")
    fun recommendThread(
        @Query("hash") formhash: String,
        @Query("tid") tid: Int
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=recommend&do=substract")
    fun unrecommendThread(
        @Query("hash") formhash: String,
        @Query("tid") tid: Int
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=buythread")
    fun getThreadPriceInfo(
        @Query("tid") tid: Int
    ): Call<BuyThreadResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=buythread&paysubmit=yes")
    fun buyThread(
        @Field("tid") tid: Int,
        @Field("formhash") formhash: String,
        @Query("handlekey") pay: String
    ): Call<BuyThreadResult>

    @POST("$DISCUZ_API_PATH?version=5&module=report&reportsubmit=true&rtype=post&inajax=1")
    fun reportPost(
        @Query("formhash") formhash: String,
        @Query("rid") pid: Int,
        @Query("report_select") option: String,
        @Query("message") message: String
    ): Call<ApiMessageActionResult>

    @get:GET("$DISCUZ_API_PATH?version=4&module=check")
    val checkResult: Call<AddCheckResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=login")
    fun loginCall(@FieldMap options: HashMap<String, String>): Call<LoginResult>

    @GET("$DISCUZ_API_PATH?version=4&module=newthreads&limit=20")
    fun newThreadsResult(
        @Query("fids") fids: String,
        @Query("start") start: Int
    ): Call<NewThreadsResult>

    @get:GET("$DISCUZ_API_PATH?version=4&module=smiley")
    val smileyResult: Call<SmileyResult>

    @GET("$DISCUZ_API_PATH?version=4&module=mypm&subop=view")
    fun getPrivateMessageListResult(
        @Query("touid") toUid: Int,
        @Query("page") pageString: String
    ): Call<PrivateMessageResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&ac=pm&op=send&daterange=0&module=sendpm&pmsubmit=yes")
    fun sendPrivateMessage(
        @Query("plid") plid: Int,
        @Query("pmid") pmid: Int,
        @Field("formhash") formHash: String,
        @Field("message") message: String,
        @Field("touid") toUid: String
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=mythread")
    fun myThreadResult(
        @Query("page") page: Int
    ): Call<DisplayThreadsResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=sendreply&action=reply&replysubmit=yes")
    fun replyThread(
        @Query("fid") fid: Int,
        @Query("tid") tid: Int,
        @FieldMap options: HashMap<String, String>
    ): Call<ApiMessageActionResult>

    @GET("$DISCUZ_API_PATH?version=4&module=mynotelist")
    fun myNotificationList(
        @Query("view") view: String?,
        @Query("type") type: String?,
        @Query("page") page: Int
    ): Call<UserNoteListResult>

    @FormUrlEncoded
    @POST("$DISCUZ_API_PATH?version=4&module=topicadmin&action=moderate&modsubmit=yes&sendreasonpm=on")
    fun adminThread(
        @Field("formhash") formhash: String,
        @Field("fid") fid: Int,
        @Field("moderate[]") tids: List<Int>,
        @Field("operations[]") operations: List<String>,
        @Field("sticklevel") stickLevel :Int,
        @Field("digestlevel") digestLevel :Int,
        @Field("reason") reason: String,
    ): Call<ApiMessageActionResult>

    companion object {
        const val DISCUZ_API_PATH = "/api/mobile/index.php"
    }
}